geniza/corpus/ja.py

```diff
@@ -79,18 +79,89 @@ def arabic_to_ja(text):
     return re.sub(re_he_final_letters, lambda m: he_final_letters[m.group(0)], text)
 
 
+# def locate_quoted_phrases(text):
+#     import re
+#     return [m.span() for m in re.finditer(r'"[^"]*"', text)]
+
+# def extract_quoted_phrases(text):
+#     return [text[m[0]:m[1]] for m in locate_quoted_phrases(text)]
+
+# def extract_arabic_phrases(text):
+#     return [
+#         phrase for phrase in extract_quoted_phrases(text) if contains_arabic(phrase)
+#     ]
+
+
+# def extract_arabic_words_outside_phrases(text):
+#     arabic_phrases = extract_arabic_phrases(text)
+
+#     # remove those arabic phrases from original string
+#     for phrase in arabic_phrases:
+#         text = text.replace(phrase, "")
+
+#     # get remaining arabic words
+#     arabic_words = extract_arabic_words(text)
+
+#     # return words
+#     return arabic_words
+
+
+# def extract_arabic_words(text):
+#     return re_AR_letters.findall(text)
+
+
+def tokenize_words_and_phrases(text):
+    import re
+
+    return [x for x in re.findall(r'"[^"]*"|[\w]+', text) if x]
```

Comment:
is there a way to adapt this regex to revise the regex I was using before so that we find and extract only arabic phrases and words? I think it's fine to limit to exact phrases that only contain arabic letters and whitespace (and possibly punctuation?); seems reasonable to me to assume that if the person searching uses other characters beyond that, they know what they're doing and we shouldn't mess with their query.

Reply:
One reason I turned to tokenizing the entire search query is that it makes replacing elements in the search easier. My original approach of using string replace failed because if an arabic word is present as a single search term as well as in a phrase, then (with the old code) it could have been replaced with a (judeo-arabic|arabic) equivalent already. Better to have a list of the elements in the search, I thought.

There's prob a way of modifying this so it only considers Arabic letters: ```r'"[\u0600-\u06FF]+"|[\u0600-\u06FF]+'```. But then we get a list of just those elements and it becomes difficult to return a replaced version of the original search query string.

Reply:
I apologize, I didn't read your regex carefully enough! I assumed it was doing something a little bit different.

What I meant to propose was that we revise the regex to this:
```python
re_AR_tokens = re.compile(r"\"[ \u0600-\u06FF]+\"|[\u0600-\u06FF]+")
```

You can then run the conversion and substitution on the entire token and insert the replacement back into the search string. And you don't have to worry about the string replace problem you describe above because arabic words inside phrases are only matched once.